### DISCUSSION TITLE
Permissions should use resource handles instead of resource names

### DISCUSSION THREAD

**Discussion by MtDalPizzol** [statamic-rad-pack/runway#211] [Feature Requests]
Hi folks! I'm digging deep into Runway this week since I'm planning to use it for a HUGE project on our company. So, I'll be sharing everything I think could be enhanced, added or fixed in the next few days. I'm also hopping to contribute to the project.

Diving into adding custom permissions to my resources, I found that the permissions uses [an "English only name" based on automatic pluralized and singularized forms of the resource name](https://github.com/duncanmcclean/runway/blob/ce1f83de3fb46618bed6982e7dec105365fb19b2/src/ServiceProvider.php#L76).

This seems like a deviation of the pattern used by Statamic itself for collections. Statamic uses the collection handles for the permissions key/id, leaving the developers free to set the permission label in whatever way they want. For instance, localizing it. It also writes the entire key/id in lowercased form.

```php
Permission::get('edit {collection} entries')->addChild(
        Permission::make('edit locked {collection} entries')
                ->label(trans('permission.edit_locked_entries'))
);
```

Runway however uses the resource name instead of the resource handle and also capitalize it. Here's an example that shows why this is problematic:

_config/runway.php_

```php
'resources' => [
        \App\Models\Author::class => [
            'name' => 'Autores', // --> This is Brazilian Portuguese and could be set using trans('resources.author.label')
            'blueprint' => 'collections.authors.author'
        ],
]
```
Now, this resource will end up generating this:

**Handle:** author
**Singular name:** Autore (This is wrong. In Brazilian Portuguese, the correct singular would be "Autor", without the "e" at the end)
**Plural name:** Autores

Some might suggest that we could simply configure the pluralizer to use the correct language, however, there's another problem: 

[Since Runway is using the resource **name**](https://github.com/duncanmcclean/runway/blob/ce1f83de3fb46618bed6982e7dec105365fb19b2/src/ServiceProvider.php#L76), the **localized version is saved as the permission key/id**, which generates sentences mixing English with other languages, like: **"View Autores"** and **"Create new Autore"**.

So, having the permissions key/id using the resource handles will be better because it allows to keep consistency AND grammatical correctness. In the example above, it will simply be **"view author"** and **"create new author"**.

Now, since there are many, many projects already using the package saving their permissions in the current way, to avoid breaking those, we could make this configurable, keeping the current way and using a custom one in case it's provided. Something like this:

_config/runway.php_

```php
'basic_permission_keys' => [
        'view' => 'view {resource}',
        'edit' => 'edit {resource}',
        'create' => 'create {resource}',
        'delete' => 'delete {resource}'
]
```

Let me know if a PR for this would be welcome.

**Comment by duncanmcclean** [comment]
Yes, I'm happy for this to be changed. We should 100% be allowing for localisation with permission labels. 

Feel free to open a PR if you have some spare time, that would be great! 

> Hi folks! I'm digging deep into Runway this week since I'm planning to use it for a HUGE project on our company. So, I'll be sharing everything I think could be enhanced, added or fixed in the next few days. I'm also hopping to contribute to the project.

Awesome! Hopefully there's not too many things to be fixed 😆
